### ISSUE TITLE
`Flop.Phoenix.table` breaks with date filter and flop 0.21.0

### ISSUE THREAD

**Issue by andreasknoepfle** [woylie/flop_phoenix#282]
Hey @woylie,

I am trying to update our code to flop >= 0.21.0 and we have a simple date filter on a schema with a `date` column.
As soon as I have a filter value for the date filter our `Flop.Phoenix.table/1` call breaks on the `path` helper when rendering the 
sorting column paths with errors like this:

```
structs expect an :id key when converting to_param or a custom implementation of the Phoenix.Param protocol (read Phoenix.Param docs for more information), got: ~D[2023-01-01]
```

The underlying `Flop` struct looks like described in the upgrade guide: 
```
%Flop{
  after: nil,
  before: nil,
  first: nil,
  last: nil,
  limit: nil,
  offset: nil,
  order_by: [:entry_date],
  order_directions: [:asc],
  page: 1,
  page_size: 14,
  decoded_cursor: nil,
  filters: [
    %Flop.Filter{field: :entry_date, op: :>=, value: nil},
    %Flop.Filter{field: :entry_date, op: :<=, value: ~D[2023-01-01]}
  ]
}
```

> To enable Flop.Phoenix to build a query string for filter parameters, the filter value must be convertible into a string via [to_string/1](https://hexdocs.pm/elixir/Kernel.html#to_string/1). If ecto_type is set to a custom Ecto type that casts values into a struct, the [String.Chars](https://hexdocs.pm/elixir/String.Chars.html) protocol must be implemented for that struct.

But our Phoenix route helper breaks with the input it gets when building the path somewhere in [build_path/3](https://github.com/woylie/flop_phoenix/blob/0.22.2/lib/flop_phoenix.ex#L1759-L1823), since ` Phoenix.Param` is not implemented by default on `Date`s. 
From the description in the upgrade guide the requirement is just implementing `to_string`, which is implemented for `Date`. 

What am I missing? Is it expected that we implement `Phoenix.Param` for things like `Date` and `DateTime`? 

For reference the schema looks like this:
```
  @derive {
    Flop.Schema,
    sortable: [
      :entry_date
    ],
    filterable: [:entry_date],
    default_order: %{
      order_by: [:entry_date],
      order_directions: [:desc]
    }
  }
  
  schema "entries" do
    field(:entry_date, :date)
    timestamps()
  end
```

TY in advance 💚

Andi

**Comment by andreasknoepfle** [comment]
I changes the phoenix route helper to a path using Phoenix VerifiedRoutes and that seems to work 🤔.

**Comment by woylie** [comment]
> I changes the phoenix route helper to a path using Phoenix VerifiedRoutes and that seems to work 🤔.

That's interesting. We've moved everything to verified routes a while ago, so I guess I didn't notice because of that. Does this fix the issue for you? If so, I'll keep the issue open, but regard it as low priority.

**Comment by woylie** [comment]
I'm actually the same error with verified routes just with:

```elixir
~p"/page?#{[d: ~D[2000-01-01]]}"
```

So I don't know why the error disappeared for you after switching to verified routes.

I'm not sure about converting everything into a string in `to_query`. The `Phoenix.Param` protocol is responsible for URL parameters, and I wouldn't want to override that. I don't know why Phoenix doesn't implement the protocol for Date, DateTime and Time structs (presumably because different applications may have different requirements for the format to use), but I think the recommendation would be to do that in your application then:

```elixir
defimpl Phoenix.Param, for: Date do
  def to_param(%Date{} = d), do: to_string(d)
end

defimpl Phoenix.Param, for: DateTime do
  def to_param(%DateTime{} = dt), do: to_string(dt)
end

defimpl Phoenix.Param, for: Time do
  def to_param(%Time{} = t), do: to_string(t)
end
```

I'll update the changelog accordingly and add a note somewhere in the docs.

**Comment by woylie** [comment]
Documentation updated in #288, changelog updated [here](https://github.com/woylie/flop/commit/c2de2df1994a4344bc112566852bee0ff91dd91f).
